File: src/main/scala/com/ajjpj/amapper/javabean/builder/JavaBeanSupport.scala
package com.ajjpj.amapper.javabean.builder

import java.lang.reflect.Method
import com.ajjpj.amapper.javabean.{JavaBeanTypes, JavaBeanType}
import com.ajjpj.amapper.javabean.JavaBeanTypes._
import scala.Some
import com.ajjpj.amapper.javabean.propbased.{MethodBasedPropertyAccessor, PropertyAccessor}
import com.ajjpj.amapper.core.{NoQualifier, AQualifier}


/**
 * @author arno
 */
object JavaBeanSupport extends JavaBeanSupport {
  def isDeferred(isPropDeferred: IsDeferredStrategy, methods: Option[Method]*) = methods.flatMap(x=>x).exists(isPropDeferred(_))
}

class JavaBeanSupport {
  //TODO configurable --> public only or all?!
  def allMethods(cls: Class[_]) = cls.getMethods

  def beanProp(cls: Class[_], name: String, isPropDeferred: IsDeferredStrategy, qualifierExtractor: QualifierExtractor) = {
    val setters = asSetterMethods(cls, name).flatMap(x => asSetter(x, isPropDeferred, qualifierExtractor))
    val getters = asGetterMethods(cls, name).flatMap(x => asGetter(x, isPropDeferred, qualifierExtractor))

    (getters.size, setters.size) match {
      case (1, 1) => Some (getters(0).copy(setter = setters(0).setter, setterType = setters(0).setterType, targetQualifier = setters(0).targetQualifier).asProperty)
      case (1, 0) => Some (getters(0).asProperty)
      case (0, 1) => Some (setters(0).asProperty)
      case _      => None
    }
  }

  def allBeanProps(cls: Class[_], isPropDeferred: IsDeferredStrategy, qualifierExtractor: QualifierExtractor): Map[String, PropertyAccessor] = {
    var raw = Map[String, AccessorDetails]() withDefault (name => AccessorDetails (name, None, None, null, null, NoQualifier, NoQualifier, isPropDeferred))

    allMethods(cls).foreach(mtd => {
      asGetter(mtd, isPropDeferred, qualifierExtractor) match {
        case Some(details) =>
          val merged = raw(details.name).copy(getter = details.getter, getterType = details.getterType, sourceQualifier = details.sourceQualifier)
          raw += (details.name -> merged)
        case None =>
      }
      asSetter(mtd, isPropDeferred, qualifierExtractor) match {
        case Some(details) =>
          //TODO check for duplicates and warn, keep only the one matching the getter
          val merged = raw(details.name).copy(setter = details.setter, setterType = details.setterType, targetQualifier = details.targetQualifier)
          raw += (details.name -> merged)
        case None =>
      }
    })

    raw.values.map(x => x.name -> x.asProperty).toMap
  }

  def asUniqueGetterMethod(cls: Class[_], propName: String): Option[Method] = asGetterMethods(cls, propName) match {
    case s if s.length == 1 => Some(s(0))
    case _ => None
  }

  def asGetterMethods(cls: Class[_], propName: String): Seq[Method] = {
    val getterNames = this.getterNames(propName)
    allMethods(cls).filter(x => getterNames.contains(x.getName)).flatMap(x => asGetter(x, null, NoQualifierExtractor)).flatMap(_.getter)
  }

  def asUniqueSetterMethod(cls: Class[_], propName: String): Option[Method] = asSetterMethods(cls, propName) match {
    case s if s.length == 1 => Some(s(0))
    case _ => None
  }

  def asSetterMethods(cls: Class[_], propName: String): Seq[Method] = {
    val setterName = this.setterName(propName)
    allMethods(cls).filter(_.getName == setterName).flatMap(x => asSetter(x, null, NoQualifierExtractor)).flatMap(_.setter)
  }

  def asGetter(mtd: Method, isPropDeferred: IsDeferredStrategy, qualifierExtractor: QualifierExtractor): Option[AccessorDetails] = mtd match {
    case _ if mtd.getDeclaringClass == classOf[java.lang.Object] => None
    case _ if mtd.getParameterTypes.size != 0 => None
    case _ if mtd.getName startsWith "get" => Some (AccessorDetails(beanPropName(mtd.getName.substring(3)), Some(mtd), None, tpe(mtd.getGenericReturnType), null, qualifierExtractor.extract(mtd), NoQualifier, isPropDeferred))
    case _ if JavaBeanTypes.normalized(mtd.getReturnType) != classOf[java.lang.Boolean] => None
    case _ if mtd.getName startsWith "is"  => Some (AccessorDetails(beanPropName(mtd.getName.substring(2)), Some(mtd), None, tpe(mtd.getGenericReturnType), null, qualifierExtractor.extract(mtd), NoQualifier, isPropDeferred))
    case _ if mtd.getName startsWith "has" => Some (AccessorDetails(beanPropName(mtd.getName.substring(3)), Some(mtd), None, tpe(mtd.getGenericReturnType), null, qualifierExtractor.extract(mtd), NoQualifier, isPropDeferred))
    case _ => None
  }

  def asSetter(mtd: Method, isPropDeferred: IsDeferredStrategy, qualifierExtractor: QualifierExtractor): Option[AccessorDetails] = mtd match {
    case _ if mtd.getName.startsWith("set") && mtd.getParameterTypes.size == 1 =>
      Some(AccessorDetails(beanPropName(mtd.getName.substring(3)), None, Some(mtd), null, tpe(mtd.getGenericParameterTypes()(0)), NoQualifier, qualifierExtractor.extract(mtd), isPropDeferred))
    case _ => None
  }

  private def toFirstUpper(s: String) = s(0).toUpper + s.substring(1)

  def setterName(propName: String) = "set" + toFirstUpper(propName)
  def getterNames(propName: String) = Set("get" + toFirstUpper(propName), "is" + toFirstUpper(propName), "has" + toFirstUpper(propName))

  def beanPropName(rawNameWithoutPrefix: String): String = rawNameWithoutPrefix match {
    case n if n.length > 1 && n.charAt(1).isUpper => n // special case defined in Java Beans standard
    case n => n(0).toLower + n.substring(1)
  }
}

case class AccessorDetails(name: String, getter: Option[Method], setter: Option[Method], getterType: JavaBeanType[_<:AnyRef], setterType: JavaBeanType[_<:AnyRef], sourceQualifier: AQualifier, targetQualifier: AQualifier, isPropDeferred: IsDeferredStrategy) {
  def isDeferred = JavaBeanSupport.isDeferred(isPropDeferred, getter, setter)
  def asProperty = MethodBasedPropertyAccessor(name, getter, setter, isDeferred, if(getter.isDefined) getterType else setterType, sourceQualifier, targetQualifier) //TODO check that they match
}

File: src/main/java/com/ajjpj/amapper/javabean/builder/JavaBeanSupport.java
package com.ajjpj.amapper.javabean.builder;

import com.ajjpj.abase.collection.immutable.AOption;
import com.ajjpj.amapper.core.tpe.AQualifier;
import com.ajjpj.amapper.javabean.JavaBeanType;
import com.ajjpj.amapper.javabean.JavaBeanTypes;
import com.ajjpj.amapper.javabean.builder.qualifier.AQualifierExtractor;
import com.ajjpj.amapper.javabean.propbased.accessors.AMethodBasedPropertyAccessor;
import com.ajjpj.amapper.javabean.propbased.accessors.APropertyAccessor;

import java.lang.reflect.Method;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;


/**
 * This is a collection of generic helper methods for working with Java Beans
 *
 * @author arno
 */
public class JavaBeanSupport {
    //TODO public or other visibilities as well? --> recursively up the inheritance, getDeclaredMethods(), setAccessible(true)

    public static final String[] GETTER_PREFIXES = new String[] {"get", "is", "has"};

    private final AIsDeferredStrategy deferredStrategy;
    private final AQualifierExtractor qualifierExtractor;

    public JavaBeanSupport(AIsDeferredStrategy deferredStrategy, AQualifierExtractor qualifierExtractor) {
        this.deferredStrategy = deferredStrategy;
        this.qualifierExtractor = qualifierExtractor;
    }

    public AOption<APropertyAccessor> getBeanProperty(Class<?> cls, String name) throws Exception {
        final AOption<AccessorDetails> optGetter = getGetter(cls, name);
        if(optGetter.isEmpty()) {
            return AOption.none();
        }

        final AccessorDetails getter = optGetter.get();
        final AOption<AccessorDetails> optSetter = getSetterFor(name, getter.method);

        if(optSetter.isDefined()) {
            final AccessorDetails setter = optSetter.get();
            return AOption.some((APropertyAccessor) new AMethodBasedPropertyAccessor(name, getter.method, setter.method, getter.isDeferred || setter.isDeferred, getter.tpe, getter.qualifier, setter.qualifier));
        }
        else {
            return AOption.some((APropertyAccessor) new AMethodBasedPropertyAccessor(name, getter.method, null, getter.isDeferred, getter.tpe, getter.qualifier, AQualifier.NO_QUALIFIER));
        }
    }

    public Map<String, APropertyAccessor> getAllProperties(Class<?> cls) throws Exception {
        final Map<String, APropertyAccessor> result = new HashMap<String, APropertyAccessor>();

        for(Method mtd: cls.getMethods()) {
            if(mtd.getParameterTypes().length > 0) {
                continue;
            }
            if(mtd.getDeclaringClass() == Object.class) {
                continue;
            }

            for(String prefix: GETTER_PREFIXES) {
                if(mtd.getName().startsWith(prefix)) {
                    final AOption<APropertyAccessor> optProp = getBeanProperty(cls, methodNameSuffixToPropName(mtd.getName().substring(prefix.length())));
                    if(optProp.isDefined()) {
                        result.put(optProp.get().getName(), optProp.get());
                    }
                }
            }
        }
        return result;
    }

    private static String propNameToMethodNameSuffix(String propName) {
        return Character.toUpperCase(propName.charAt(0)) + propName.substring(1);
    }

    private static String methodNameSuffixToPropName(String methodNameSuffix) {
        if(methodNameSuffix.length() > 1 && Character.isUpperCase(methodNameSuffix.charAt(1))) {
            return methodNameSuffix;
        }
        return Character.toLowerCase(methodNameSuffix.charAt(0)) + methodNameSuffix.substring(1);
    }

    private Set<AccessorDetails> getGetters(Class<?> cls, String name) throws Exception {
        final String methodNameSuffix = propNameToMethodNameSuffix(name);
        final Set<AccessorDetails> result = new HashSet<AccessorDetails>();

        for(String prefix: GETTER_PREFIXES) {
            try {
                final Method getter = cls.getMethod(prefix + methodNameSuffix);
                result.add(new AccessorDetails(getter, name, JavaBeanTypes.create(getter.getGenericReturnType()), qualifierExtractor.extract(getter), deferredStrategy.isDeferred(getter)));
            } catch (Exception e) { //
            }
        }

        return result;
    }

    public AOption<AccessorDetails> getGetter(Class<?> cls, String name) throws Exception {
        final Set<AccessorDetails> all = getGetters(cls, name);
        switch(all.size()) {
            case 0: return AOption.none();
            case 1: return AOption.some(all.iterator().next());
            default: return AOption.none(); //TODO log non-unique getter
        }
    }

    public AOption<AccessorDetails> getSetterFor(String name, Method getter) throws Exception {
        // passing in the getter instead of the JavaBeanType takes care of the whole primitive / boxed dichotomy
        try {
            final Method setter = getter.getDeclaringClass().getMethod("set" + propNameToMethodNameSuffix(name), getter.getReturnType());
            return AOption.some(new AccessorDetails(setter, name, JavaBeanTypes.create(getter.getGenericReturnType()), qualifierExtractor.extract(setter), deferredStrategy.isDeferred(setter)));
        }
        catch(Exception exc) { //
            return AOption.none();
        }
    }

    public static class AccessorDetails {
        public final Method method;
        public final String propName;
        public final JavaBeanType<?> tpe;
        public final AQualifier qualifier;
        public final boolean isDeferred;

        public AccessorDetails(Method method, String propName, JavaBeanType<?> tpe, AQualifier qualifier, boolean deferred) {
            this.method = method;
            this.propName = propName;
            this.tpe = tpe;
            this.qualifier = qualifier;
            isDeferred = deferred;
        }

        public boolean isGetter() {
            return method.getParameterTypes().length == 0;
        }
    }
}
